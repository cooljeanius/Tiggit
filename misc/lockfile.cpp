#include "lockfile.hpp"

#include <boost/filesystem.hpp>
#include <fstream>
#include <assert.h>

#ifdef _WIN32
#include "windows.h"
int getPID() { return GetCurrentProcessId(); }
#else
#ifdef __unix
#include <unistd.h>
int getPID() { return getpid(); }
#endif
#endif

using namespace Misc;
using namespace boost::filesystem;
using namespace std;

int LockFile::counter = 0;

bool LockFile::lock(bool force)
{
  assert(!locked);
  assert(file != "");

  int lockValue = (getPID()&0xffffff) | (ID << 24);

  // Is the resource already locked?
  if(exists(file) && !force)
    return false;

  // Write our lock value
  {
    ofstream ofs(file.c_str(), ios::binary);

    // Could we create the lock file?
    if(!ofs)
      return false;

    ofs.write((char*)&lockValue, 4);
  }

  // Confirm our lock value
  ifstream ifs(file.c_str(), ios::binary);
  if(!ifs)
    return false;

  int rval;
  ifs.read((char*)&rval, 4);

  locked = (rval == lockValue);
  return locked;
}

void LockFile::unlock()
{
  if(locked)
    {
      remove(file);
      locked = false;
    }
}
