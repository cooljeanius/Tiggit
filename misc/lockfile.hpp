#ifndef __LOCKFILE_HPP_
#define __LOCKFILE_HPP_

#include <string>

/* A lockfile is used to secure sole ownership of a resource. It's
   used to lock repositories so multiple processes won't manipulate it
   at the same time.
 */

namespace Misc
{
  struct LockFile
  {
    LockFile(const std::string &_file = "")
      : file(_file), locked(false)
    { ID = counter++; }
    ~LockFile() { unlock(); }

    /* Lock function. Returns false if the lock failed.

       In case of a stale lock file (the dead remains of a terminated
       process that failed to clean up after itself), you can attempt
       to force the lock by ignoring existing files. Use with caution.
     */
    bool lock(bool force=false);
    void unlock();
    bool lock(const std::string &file, bool force=false)
    { setFile(file); return lock(force); }

    void setFile(const std::string &_file)
    { unlock(); file = _file; }
    bool isLocked() const { return locked; }

  private:

    std::string file;
    bool locked;
    int ID;

    static int counter;
  };
}

#endif
