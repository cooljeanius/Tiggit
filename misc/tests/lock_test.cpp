#include "lockfile.hpp"

#include <iostream>
using namespace std;

void w(bool b)
{
  if(b) cout << "YES\n";
  else cout << "NO\n";
}

int main()
{
  Misc::LockFile lockA("_lock1"), lockB;
  lockB.setFile("_lock1");

  cout << "Locking B:\n";
  w(lockB.lock());
  w(lockA.lock());

  cout << "Checking:\n";
  w(lockA.isLocked());
  w(lockB.isLocked());

  cout << "Aquiring A normally:\n";
  w(lockA.lock());
  lockB.unlock();
  w(lockA.lock());

  cout << "Checking:\n";
  w(lockA.isLocked());
  w(lockB.isLocked());

  cout << "Aquiring B by force:\n";
  w(lockB.lock());
  w(lockB.lock(true));

  cout << "Checking:\n";
  w(lockA.isLocked());
  w(lockB.isLocked());
  return 0;
}
