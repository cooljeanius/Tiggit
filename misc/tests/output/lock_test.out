Locking B:
YES
NO
Checking:
NO
YES
Aquiring A normally:
NO
YES
Checking:
YES
NO
Aquiring B by force:
NO
YES
Checking:
YES
YES
